modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PendingRows.java

```diff
@@ -81,8 +77,12 @@ public void addPendingRowIds(UUID txId, Collection<RowId> rowIds) {
      *
      * @param txId Transaction ID.
      */
-    public void removePendingRowIds(UUID txId) {
-        txsPendingRowIds.remove(txId);
+    public void removePendingRowIds(UUID txId, Set<RowId> pendingRowIds) {
```

Comment:
It's not a problem. First of all generally speaking lazy writeIntent resolution gives us correctness. Cleanup on tx finish is an optimization, very important one, but still optimization. We don't expect such crashes to be an often case, so that we might prefer pimplier API to rare performance related inefficiency.

Reply:
If RW transaction touches a writeIntent it'll do the cleanup before any further activities. It's a mandatory action, however it's not a cleanup on tx finish that I've mentioned previously, meaning that tx finish cleanup still should be considered as an optimization. It's absolutely required for a RW transaction to cleanup specific writeIntent that was touched, however in addition to that, as we previously discussed, it may try to cleanup all local writeIntents with same txId. 
All in all, I still believe that removeAndGet should work here.